File: engines/pink/objects/carchive.cpp
#include "carchive.h"

#include "cobject.h"
#include "cruntimeclass.h"

#include "common/file.h"
#include "common/str.h"
#include "common/array.h"

#include "cgame.h"
#include "cnamedobject.h"
#include "cmoduleproxy.h"
#include "cobject.h"

namespace Pink {
	CArchive::CArchive(Common::File *file) : _known_items_length(1) {
		_file = file;
	}
	
	CArchive::~CArchive() {
	}

	void CArchive::mapObject(CObject *obj) {
		_known_objects[_known_items_length++] = obj;
	}

	void CArchive::mapClass(CRuntimeClass *runtimeclass) {
		_known_classes[_known_items_length++] = runtimeclass;
	}

	uint8 CArchive::readBYTE() {
		return _file->readByte();
	}

	uint16 CArchive::readWORD() {
		return _file->readUint16LE();
	}

	uint32 CArchive::readDWORD() {
		return _file->readUint32LE();
	}

	Common::String *CArchive::readCString() {
		uint32 length = readBYTE();
		if (length == 0xff)
			length = readWORD();
		if (length == 0xffff)
			length = readDWORD();

		char *tmpstr = new char[length];
		_file->read(tmpstr, length);
		Common::String *str = new Common::String(tmpstr, length);
		delete[] tmpstr;

		return str;
	}

	CObject *CArchive::readCObject() {
		uint16 wTag = _file->readUint16LE();
		uint32 tag;
		if (wTag == 0x7fff) // wide tag
			tag = _file->readUint32LE();
		else
			tag = ((wTag & 0x8000) << 16) | (wTag & 0x7fff);

		if (tag & 0x80000000) { // new object
			CRuntimeClass *runtimeclass;
			if (wTag == 0xffff) { // new class
				uint16 schemaNum = _file->readUint16LE();
				uint16 classnameLength = _file->readUint16LE();
				if (classnameLength > 64)
					return NULL;

				char *classname = new char[classnameLength + 1];
				_file->read(classname, classnameLength);
				classname[classnameLength] = 0;
				runtimeclass = CRuntimeClass::getRuntimeClassByName(classname);
				if (!runtimeclass) {
					warning("PINK: Could not find class named \"%s\"", classname);
					return NULL;
				}
				mapClass(runtimeclass);
			} else {
				uint32 index = tag & 0x7fffffff;
				if (_known_items_length > index)
					runtimeclass = _known_classes[index];
				else
					return NULL;
			}
			CObject *obj = runtimeclass->createObject();
			mapObject(obj);
			obj->deserialize(*this);
			return obj;
		} else {
			if (_known_items_length > tag)
				return _known_objects[tag];
			else
				return NULL;
		}
	}

	Common::Array<CObject *> *CArchive::readCObArray() {
		return readCObArray(NULL);
	}

	Common::Array<CObject *> *CArchive::readCObArray(CRuntimeClass *runtimeclass) {
		Common::Array<CObject *> *arr = new Common::Array<CObject *>();
		readCObArray(*arr);
		return arr;
	}

	Common::Array<CObject *> &CArchive::readCObArray(Common::Array<CObject *> &arr) {
		return readCObArray(arr, NULL);
	}

	Common::Array<CObject *> &CArchive::readCObArray(Common::Array<CObject *> &arr, CRuntimeClass *runtimeclass) {
		uint16 length = _file->readUint16LE();
		CObject **tmparray = (CObject **)calloc(length, sizeof(CObject *));
		for (uint16 i = 0; i < length; i++) {
			CObject *obj = readCObject();
			if (!obj || runtimeclass && !runtimeclass->isInstance(obj)) {
				if (obj)
					delete obj;

				for (uint16 j = 0; j < i; j++) {
					delete tmparray[j];
				}
				free(tmparray);
				return arr;
			} else {
				tmparray[i] = obj;
			}
		}
		arr = Common::Array<CObject *>(tmparray, length);
		free(tmparray);
		return arr;
	}

	Common::Array<Common::String *> *CArchive::readCStringArray() {
		uint16 length = _file->readUint16LE();
		Common::String **tmparray = (Common::String **)calloc(length, sizeof(Common::String *));
		for (uint16 i = 0; i < length; i++) {
			Common::String *obj = readCString();
			tmparray[i] = obj;
		}
		Common::Array<Common::String *> *arr = new Common::Array<Common::String *>(tmparray, length);
		free(tmparray);
		return arr;
	}

};

File: engines/pink/OrbFile/CArchive.py
import struct
import new

from CObject import *

def logged(func):
    return func

    import functools

    @functools.wraps(func)
    def a(*args, **kwargs):
        ret = func(*args, **kwargs)
        print ret
        return ret
    return a

class CArchive(object):
    _serializable = {}

    @classmethod
    def register(cls, clsObj):
        cls._serializable[clsObj.__name__] = clsObj
        #print len(cls._serializable), clsObj.__name__
    
    def __init__(self, f):
        self._f = f
        self._map = []
        self._objTypes = []
        self.mapObject(None)

    def mapObject(self, obj):
        self._insertObj(obj)

    def _insertObj(self, obj):
        self._map.append(obj)
        self._objTypes.append(2)

    def _insertClass(self, cls):
        self._map.append(cls)
        self._objTypes.append(1)

    def readUInt(self):
        (num,) = struct.unpack('I', self._f.read(4))
        return num

    def readInt(self):
        (num,) = struct.unpack('i', self._f.read(4))
        return num

    def readUShort(self):
        (num,) = struct.unpack('H', self._f.read(2))
        return num

    def readShort(self):
        (num,) = struct.unpack('h', self._f.read(2))
        return num

    def readUByte(self):
        (num,) = struct.unpack('B', self._f.read(1))
        return num

    def readByte(self):
        (num,) = struct.unpack('b', self._f.read(1))
        return num

    def readCString(self):
        size = self.readUByte()
        if size == 0xff:
            raise TypeError('size is -1')
        text = self._f.read(size)
        return text

    def readCount(self):
        count = self.readUShort()
        if count == 0xff:
            raise TypeError('count is -1')
        return count

    def readCStringArray(self):
        length = self.readCount()
        if length == 0xff:
            raise TypeError('length is -1')
        arr = []
        for i in xrange(length):
            s = self.readCString()
            arr.append(s)
        return arr

    def readCObArray(self):
        length = self.readCount()
        arr = []
        for i in xrange(length):
            obj = self.readObject()
            if obj is None:
                arr += [None] * (length - i)
                break
            arr.append(obj)
        return arr

    @logged
    def readObject(self, baseClass = CObject):
        wTag = self.readUShort()
        if wTag == 0x7fff: # big object tag
            tag = self.readUInt()
        else:
            tag = ((wTag & 0x8000) << 16) | (wTag & 0x7fff)

        if (tag & 0x80000000) == 0:
            try:
                obj = self._map[tag]
            except IndexError:
                raise IndexError("Undefined tag %d" % tag)
            if isinstance(obj, baseClass):
                return obj
            raise Exception("Not the right object!")
        
        if wTag == 0xffff: # new class tag
            schemaNum = self.readUShort()
            classname_length = self.readUShort()
            if classname_length > 64:
                return None
            
            classname = self._f.read(classname_length)
            try:
                cls = CArchive._serializable[classname]
            except:
                raise TypeError("Unknown class %s" % classname) 

            self._insertClass(cls)
        else:
            index = tag & 0x7fffffff
            if len(self._map) > index:
                cls = self._map[index]
            else:
                return None

        if not issubclass(cls, baseClass):
            raise TypeError("Expected a %s object, got %s" % (baseClass.__name__, cls.__name__))
        
        obj = cls()
        self._insertObj(obj)
        obj.deserialize(self)
        return obj

__all__ = ["CArchive"]
